{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027c2abd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: generics\n",
      "\n",
      "\n",
      "Attaching package: ‘generics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
      "    setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
      "    unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'zellkonverter':\n",
      "  method                                             from      \n",
      "  py_to_r.pandas.core.arrays.categorical.Categorical reticulate\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘celda’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    params\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SingleCellExperiment); library(DropletUtils); library(zellkonverter); library(celda)\n",
    "\n",
    "# Load raw and set barcodes\n",
    "sce_raw_filtered <- DropletUtils::read10xCounts(\"00_data/raw.h5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76d43729",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 61313 rows and 3 columns\n",
       "                          ID                 Symbol            Type\n",
       "                 <character>            <character>     <character>\n",
       "1         ENSSSCG00000028996                ALDH1A1 Gene Expression\n",
       "2         ENSSSCG00000005267                  ANXA1 Gene Expression\n",
       "3         ENSSSCG00000005268                   RORB Gene Expression\n",
       "4         ENSSSCG00000005269                  TRPM6 Gene Expression\n",
       "5         ENSSSCG00000031382                C9orf40 Gene Expression\n",
       "...                      ...                    ...             ...\n",
       "61309 AEMK02000695.1:5684-.. AEMK02000695.1:5684-..           Peaks\n",
       "61310 AEMK02000695.1:9057-.. AEMK02000695.1:9057-..           Peaks\n",
       "61311 AEMK02000695.1:11026.. AEMK02000695.1:11026..           Peaks\n",
       "61312 AEMK02000695.1:12452.. AEMK02000695.1:12452..           Peaks\n",
       "61313 AEMK02000695.1:13456.. AEMK02000695.1:13456..           Peaks"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sce_raw_filtered@rowRanges@elementMetadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cb2291c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sce_raw_filtered_filtered <- sce_raw_filtered[sce_raw_filtered@rowRanges@elementMetadata$Type == \"Gene Expression\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec1abcc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 22040 722431 \n",
       "metadata(1): Samples\n",
       "assays(1): counts\n",
       "rownames(22040): ENSSSCG00000028996 ENSSSCG00000005267 ...\n",
       "  ENSSSCG00000018094 ENSSSCG00000034846\n",
       "rowData names(3): ID Symbol Type\n",
       "colnames: NULL\n",
       "colData names(2): Sample Barcode\n",
       "reducedDimNames(0):\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sce_raw_filtered_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "31af49d2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Wed Sep  3 17:12:02 2025 ..  12315  cells in the background matrix were removed as they were found in  the filtered matrix.\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Wed Sep  3 17:12:03 2025 .. Analyzing all cells\n",
      "\n",
      "Wed Sep  3 17:12:03 2025 .... Converting to sparse matrix\n",
      "\n",
      "Wed Sep  3 17:13:18 2025 .... Generating UMAP\n",
      "\n",
      "Wed Sep  3 17:13:50 2025 .... Estimating contamination\n",
      "\n",
      "Wed Sep  3 17:13:51 2025 ...... Completed iteration: 10 | converge: 0.02596\n",
      "\n",
      "Wed Sep  3 17:13:52 2025 ...... Completed iteration: 20 | converge: 0.005476\n",
      "\n",
      "Wed Sep  3 17:13:52 2025 ...... Completed iteration: 30 | converge: 0.001876\n",
      "\n",
      "Wed Sep  3 17:13:53 2025 ...... Completed iteration: 38 | converge: 0.0009808\n",
      "\n",
      "Wed Sep  3 17:13:53 2025 .. Calculating final decontaminated matrix\n",
      "\n",
      "Wed Sep  3 17:13:54 2025 .. Converting decontaminated matrix to DelayedMatrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 1.855598 mins\n",
      "\n",
      "--------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (is.null(colnames(sce_raw_filtered)) || length(colnames(sce_raw_filtered))==0)\n",
    "  colnames(sce_raw_filtered) <- as.character(SummarizedExperiment::colData(sce_raw_filtered)[[\"Barcode\"]])\n",
    "\n",
    "# Foreground barcodes\n",
    "ad_fg <- zellkonverter::readH5AD(\"00_data/foreground.h5ad\")\n",
    "norm <- function(x) sub(\"-1$\",\"\", x)\n",
    "keep_fg <- norm(colnames(sce_raw_filtered)) %in% norm(colnames(ad_fg))\n",
    "sce_fg <- sce_raw_filtered[, keep_fg, drop=FALSE]\n",
    "\n",
    "# Read cluster_type produced in Python\n",
    "lab <- read.csv(\"51_annot_cells_types.csv\", row.names=1, check.names = FALSE)\n",
    "m <- match(norm(colnames(sce_fg)), norm(rownames(lab)))\n",
    "sce_fg$cluster_annotation <- factor(lab$cluster_annotation[m])\n",
    "\n",
    "if (any(is.na(sce_fg$cluster_annotation))) stop(\"Some barcodes did not match labels. Check CSV/barcodes.\")\n",
    "\n",
    "set.seed(42)\n",
    "sce_out <- celda::decontX(sce_fg, ,background = sce_raw_filtered, z = \"cluster_annotation\")\n",
    "\n",
    "# Save\n",
    "dir.create(\"53_results_clusters\", showWarnings=FALSE)\n",
    "HDF5Array::saveHDF5SummarizedExperiment(sce_out, dir=\"53_results_clusters/decontx_with_clusters_SCEh5\", replace=TRUE)\n",
    "zellkonverter::writeH5AD(sce_out, \"53_results_clusters/decontx_with_clusters.h5ad\", X_name=\"decontXcounts\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
